File: src/300_Longest_Increasing_Subsequence/main.py
from typing import List

class Solution:
    # Runtime: 96 ms, faster than 75.75%
    # Memory Usage: 14.8 MB, less than 5.96%
    def lengthOfLIS(self, nums: List[int]) -> int:
        lenByNow = []
        for i in range(len(nums)):
            # 二分查找复习一下，找到递增序列lenByNow中，小于且最接近num的那个值的index
            num = nums[i]
            lhs, rhs = 0, len(lenByNow)
            while lhs < rhs:
                # mid = lhs + (rhs-lhs) 怎么大意成这样！！！
                mid = lhs + int((rhs-lhs)/2)
                if lenByNow[mid] == num:
                    rhs = mid
                elif lenByNow[mid] > num:
                    rhs = mid
                else:
                    lhs = mid+1
            
            # lhs是需要的index
            if lhs == len(lenByNow):
                lenByNow.append(num)
            else:
                lenByNow[lhs] = num
        # 因为不需要返回具体的序列，所以直接是len(lenByNow)就OK了
        return len(lenByNow)

    # Runtime: 3928 ms, faster than 34.53%
    # Memory Usage: 14.4 MB, less than 92.51% 
    def lengthOfLIS(self, nums: List[int]) -> int:
        # 这里dp的定义很巧妙，dp[i]保存以nums[i]为结尾的subsequence的最长值，这样才能具有子问题属性
        size = len(nums)
        lenByNow = [1 for _ in range(size)]
        
        for i in range(1,size):
            # 然后内部从i-1到0
            for j in range(i-1,-1,-1):
                if nums[i] > nums[j]: # 题目要求是strick严格大于
                    # lenByNow[i] = lenByNow[j]+1
                    # break #这里greedy了一下，肯定越靠近i的越长
                    # ——不对！不能break要继续努力，前面可能还有更好的
                    lenByNow[i] = max(lenByNow[i], lenByNow[j]+1)

        # 由于dp不是最终的目的，而是中间过程，最后还要遍历一遍dp，找到里面的最值
        rtn = max(lenByNow)
        return rtn
        
if __name__ == "__main__":
  gua = Solution()
  rtn = gua.lengthOfLIS(nums = [10,9,2,5,3,7,101,18])
  print("expect result = 4, actual result = %d " % rtn )

  rtn = gua.lengthOfLIS(nums = [0,1,0,3,2,3])
  print("expect result = 4, actual result = %d " % rtn )

  rtn = gua.lengthOfLIS(nums = [3,3,3,3,3,3])
  print("expect result = 1, actual result = %d " % rtn )

File: src/300_Longest_Increasing_Subsequence/Main.java

class Solution {
    //Runtime: 56 ms, faster than 61.32%
    //Memory Usage: 38.8 MB, less than 55.90%
    public int lengthOfLIS(int[] nums) {
        if (nums == null || nums.length == 0) return 0;
        
        int rtn = 0;
        int size = nums.length;
        int[] dp = new int[size];
        for (int i = 0; i < size; ++i)
            dp[i] = 1;//初始化：每个subsequence仅含一个数，故长度为1
        
        // 然后本质上是个DFS+memo的操作，这里dp的含义是从左边开头、以当前位置i结束的数组中的最长子序列
        // 从右边的计算会利用左边计算的结果，左边的结果是已经计算出来的最优解，然后最大值在过程中产生
        for (int i = 0; i < size; ++i) {
            for (int j = i-1; j >= 0; --j) {
                if (nums[i] > nums[j]) {
                    dp[i] = Math.max(dp[i], dp[j] + 1);
                }
            }
            rtn = Math.max(rtn, dp[i]);
        }
        
        return rtn;
    }
}

public class Main {
  public static void main(String[] args) {
    Solution gua = new Solution();
    int rtn;

    int [] nums = {10,9,2,5,3,7,101,18};
    rtn = gua.lengthOfLIS(nums);
    System.out.println("expect value = 4, actual value = " +rtn);
    
    nums = new int[]{0,1,0,3,2,3}; 
    rtn = gua.lengthOfLIS(nums);
    System.out.println("expect value = 4, actual value = " +rtn);
   
    nums = new int[]{1,1,1,1,1,1}; 
    rtn = gua.lengthOfLIS(nums);
    System.out.println("expect value = 1, actual value = " +rtn);
  }
}


